### ISSUE TITLE
Incorrect currying

### ISSUE THREAD

**Issue by stkb** [fable-compiler/Fable#2116]
I've had a bit of trouble with bugs in currying/uncurrying lately, but have managed to find a minimal example for one:

[Repl link](https://fable.io/repl/#?code=DYUwLgBA9gDmCWUB2AxJEC8EDKUC2IEAFAGYCu6AHhAJ4QC0AfBNQNS0CUAsAFC+iRg8AM6QsAGRFgAdAGMAFlCjDC8ACYQA2rATI0AXV68YAJ3hIwJdACJJoiKCQBzMPIBcEAKRrrxOzMcXeQcpbh5Tc0sbAEEkYQB3EBMvHz8paXkQAEMNIXsAZggAFg4IAHoyiABRShgQWUgAdl4gA&html=Q&css=Q)

``` fsharp
let optionFn = Some (fun x y -> x + y)

let list = List.choose id [optionFn]

printfn "List length: %d" (List.length list)
printfn "Answer %d" (List.head list 3 4) // Expect 7 but get 0
```

Javascript:
``` js
import { curry } from "fable-library/Util.js";
import { List } from "fable-library/Types.js";
import { head, length, choose } from "fable-library/List.js";
import { toConsole, printf } from "fable-library/String.js";
export function optionFn(x) {
  return function (y) {
    return x + y;
  };
}
export const list = choose(function (x$$1) {
  return curry(2, x$$1);
}, new List(optionFn, new List()));

(function () {
  const arg10 = length(list) | 0;
  const clo1 = toConsole(printf("List length: %d"));
  clo1(arg10);
})();

(function () {
  const arg10$$1 = head(list)(3)(4) | 0;
  const clo1$$1 = toConsole(printf("Answer %d"));
  clo1$$1(arg10$$1);
})();
```

In the javascript, `curry(2, x$$1)` is called on the input function, but it's already curried, and the end result is invalid.

I'm not sure what the essential elements are but that the function is wrapped in an Option seems to be one.

**Comment by alfonsogarciacaro** [comment]
Thanks a lot for the code to reproduce @stkb! Yes, the fact that Fable erases option creates some trouble with currying/uncurrying but hopefully we can fix it thanks to your sample 👍

**Comment by alfonsogarciacaro** [comment]
This took a while but should be finally fixed in 3.1.3 😸

**Comment by stkb** [comment]
That's great, thanks!
